finagle-core/src/main/scala/com/twitter/finagle/tracing/Id.scala

```diff
@@ -84,26 +97,30 @@ object TraceId {
         traceId.flags.setFlag(Flags.SamplingKnown)
     }
 
-    val bytes = new Array[Byte](32)
-    ByteArrays.put64be(bytes, 0, traceId.spanId.toLong)
-    ByteArrays.put64be(bytes, 8, traceId.parentId.toLong)
-    ByteArrays.put64be(bytes, 16, traceId.traceId.toLong)
+    val bytes = new Array[Byte](40)
+    ByteArrays.put64be(bytes, 0, traceId.spanId.self)
+    ByteArrays.put64be(bytes, 8, traceId.parentId.self)
+    ByteArrays.put64be(bytes, 16, traceId.traceId.self)
     ByteArrays.put64be(bytes, 24, flags.toLong)
+    if(traceId.traceIdHigh.isDefined) ByteArrays.put64be(bytes, 32, traceId.traceIdHigh.get.self)
     bytes
   }
 
   /**
    * Deserialize a TraceId from an array of bytes.
    */
   def deserialize(bytes: Array[Byte]): Try[TraceId] = {
-    if (bytes.length != 32) {
-      Throw(new IllegalArgumentException("Expected 32 bytes"))
+    // Allows for 64-bit or 128-bit trace identifiers
+    if (bytes.length != 32 && bytes.length != 40) {
+      Throw(new IllegalArgumentException("Expected 32 or 40 bytes"))
     } else {
       val span64 = ByteArrays.get64be(bytes, 0)
       val parent64 = ByteArrays.get64be(bytes, 8)
       val trace64 = ByteArrays.get64be(bytes, 16)
       val flags64 = ByteArrays.get64be(bytes, 24)
 
+      val traceIdHigh = if(bytes.length == 40) ByteArrays.get64be(bytes, 32) else 0L
```

Comment:
Why not `Option[SpanId]` instead of `Long`? Looks like your putting a `Some(SpanId(0L))` into the `TraceId.apply` below due to this (same in Trace.scala). Is this the only situations where you would expect to find a `SpanId` of 0? If so, lets do `_traceIdHigh -> traceIdHigh` and remove the method: I know that case class is pretty hairy at this point but I'd prefer the generated `unapply` method to yield as accurate results as possible.

Reply:
The reason I did it this way with the special cased 0L, is that systems requiring 128bit traces may be padded with zeros. On my team at Expedia, we hacked the 64bit trace ids to fit our tracing system by transforming the traces generated by Finagle into a UUID with zeroed most significant bits (example: `new java.util.UUID(0,traceId)`). If the special case is dropped here and not cleaned up via the `traceId` method, I assumed that would cause problems with allowing a traceIdHigh equal to 0L.

I may have misunderstood what @adriancole said with 0L being invalid.  To clarify, my understanding was that both of these are invalid:
`0000 0000 0000 0000 0000 0000 0000 0000` 
`0000 0000 0000 0000 7fff ffff ffff ffff`

Is the second case a valid 128bit identifier? If it is valid, then I agree we should just take the bits as they come and forward them along.

It's a little different from the tracing system I'll be integrating with, in which the identifiers are string representations of UUIDs with no constraint on the high 64bits (see [haystack-idl](https://github.com/ExpediaDotCom/haystack-idl/blob/4138a0660d060743d7240bf275a28046b587fc1a/proto/span.proto#L27)).